src/bvh/ordered_traverse.rs

```diff
@@ -0,0 +1,269 @@
+            bvh,
+            ray,
+            shapes,
+            stack: [(0, RestChild::None); 32],
+            node_index: 0,
+            stack_size: 0,
+            has_node: true,
+        }
+    }
+
+    /// Test if stack is empty.
+    fn is_stack_empty(&self) -> bool {
+        self.stack_size == 0
+    }
+
+    /// Push node onto stack.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `stack[stack_size]` is out of bounds.
+    fn stack_push(&mut self, nodes: (usize, RestChild)) {
+        self.stack[self.stack_size] = nodes;
+        self.stack_size += 1;
+    }
+
+    /// Pop the stack and return the node.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `stack_size` underflows.
+    fn stack_pop(&mut self) -> (usize, RestChild) {
+        self.stack_size -= 1;
+        self.stack[self.stack_size]
+    }
+
+    /// Attempt to move to the nearest to ray node child of the current node.
+    /// If it is a leaf, or the [`Ray`] does not intersect the any node [`Aabb`], `has_node` will become false.
+    fn move_nearest(&mut self) -> (usize, RestChild) {
+        let current_node_index = self.node_index;
+        match self.bvh.nodes[current_node_index] {
+            BvhNode::Node {
+                child_l_index,
+                ref child_l_aabb,
+                child_r_index,
+                ref child_r_aabb,
+                ..
+            } => {
+                
+                let (left_dist, _) = self.ray.intersection_slice_for_aabb(child_l_aabb);
+                let (right_dist, _) = self.ray.intersection_slice_for_aabb(child_r_aabb);
+
+                if left_dist > right_dist {
+                    // right is near than left
+                    self.has_node = true;
+                    self.node_index = child_r_index;
+                    let rest_child = if left_dist.is_finite() { RestChild::Left } else { RestChild::None };
+                    return (current_node_index, rest_child)
+
+                } else if left_dist.is_finite() {
+                    // left is near than right
+                    self.has_node = true;
+                    self.node_index = child_l_index;
+                    let rest_child = if right_dist.is_finite() { RestChild::Right } else { RestChild::None };
+                    return (current_node_index, rest_child)
+
+                } else {
+                    // no intersections with nearest child
+                    self.has_node = false;
+                    return (current_node_index, RestChild::None)
+                }
+            }
+            BvhNode::Leaf { .. } => {
+                self.has_node = false;
+                return (current_node_index, RestChild::None)
+            }
+        }
+    }
+
+    /// Attempt to move to the farthest to ray node child of the current node.
+    /// If it is a leaf, or the [`Ray`] does not intersect the node [`Aabb`], `has_node` will become false.
+    fn move_farthest(&mut self, rest_child: RestChild) {
```

Comment:
```suggestion
    fn move_furthest(&mut self, rest_child: RestChild) {
```
Nit: English is weird haha